llama.cpp

```diff
@@ -2836,10 +2860,103 @@ struct llama_model_loader {
+        }
+
+        uint16_t n_split = 0;
+        get_key(llm_kv(LLM_KV_SPLIT_COUNT), n_split, false);
+
+        // Build virtual GGUF/GGML contexts to represent all tensors across files
+        if (n_split > 1) {
+            uint16_t idx = 0;
+            get_key(llm_kv(LLM_KV_SPLIT_NO), idx);
+            if (idx != 0) {
+                throw std::runtime_error(format("illegal split file: %d, model must be loaded with the first split", idx));
+            }
+            get_key(llm_kv(LLM_KV_SPLIT_TENSORS_COUNT), n_tensors);
+
+            char split_prefix[4096] = {0};
+            if (!llama_split_prefix(split_prefix, fname.c_str(), fname.size(), idx, n_split)) {
+                throw std::runtime_error(format("invalid split file: %s", fname.c_str()));
+            }
+
+            size_t mem_size = n_tensors*ggml_tensor_overhead();
+            struct ggml_init_params pdata = {
+                /*.mem_size   = */ mem_size,
+                /*.mem_buffer = */ NULL,
+                /*.no_alloc   = */ true,
+            };
+
+            auto * new_ctx_meta = ggml_init(pdata);
+
+            if (trace > 0) {
+                LLAMA_LOG_INFO("%s: loading additional %d GGUFs\n", __func__, n_split);
+            }
+
+            for (; idx < n_split; idx++) {
+                char split_path[PATH_MAX] = {0};
+                struct ggml_context * split_ctx_meta = NULL;
+                struct gguf_context * split_ctx_gguf = NULL;
+                if (idx == 0) {
+                    split_ctx_gguf = ctx_gguf;
+                    split_ctx_meta = ctx_meta;
+                    strcpy(split_path, fname.c_str());
+                } else {
+                    llama_split_path(split_path, sizeof(split_path), split_prefix, idx, n_split);
+                    struct gguf_init_params split_params = {
+                        /*.no_alloc = */ true,
+                        /*.ctx      = */ &split_ctx_meta,
+                    };
+                    split_ctx_gguf = gguf_init_from_file(split_path, split_params);
+                    if (!split_ctx_gguf) {
+                        throw std::runtime_error(format("%s: failed to load GGUF split from %s\n", __func__, fname.c_str()));
+                    }
+                }
+
+                bool ok = true;
+                for (ggml_tensor * tensor = ggml_get_first_tensor(split_ctx_meta); tensor; tensor = ggml_get_next_tensor(split_ctx_meta, tensor)) {
+                    struct ggml_tensor * copy = ggml_new_tensor(new_ctx_meta, tensor->type, ggml_n_dims(tensor), tensor->ne);
+                    ok = ok && copy != NULL;
+
+                    if (!ok) {
+                        break;
+                    }
+
+                    ggml_set_name(copy, tensor->name);
+
+                    // Add the tensor to the main gguf context if not already present
+                    if (idx > 0) {
+                        gguf_add_tensor(ctx_gguf, copy);
```

Comment:
As said on #6000, this is the reason why I don't feel like this is a good idea.

If you read the source code of `gguf_add_tensor()`, you will see that `gguf_context` already manages offset of tensors internally. The current approach is quite a "patch" to me, since now the code only cares about `llama_tensor_offset`, that defeat the whole point of having `ctx_gguf` in the first place (which is to store the list of tensors and their offsets).

I understand that here you want to build a `ctx_gguf` that have a list of all tensor (maybe to be contiguous?) but the fact is that, tensor does not need to be contiguous at all, and tensors can come from different `ctx_gguf`.

So in the end I still lean to the idea of having `std::map<tensor name, ctx_gguf>`. Maybe @ggerganov or @slaren will have other opinions than mine..

Reply:
You are mixing metadata and data - no contiguous stuff here. We want tensors metadata at one place, because this is from there layers are built and backend buffers initialized.

Keeping `ctx_gguf` of each file will dupplicate tensors metadata but we only need the tensors offset in the original file.

Additionnaly, we need to know to which file/mmap the tensor data need to be loaded from, so we need to introduce a vector to map the tensor name to the file index, and I believe also to include the original offset in this vector as a simple struct does not hurt instead of keeping the whole split metadata for nothing.

Reply:
I agree that adding the tensors to a single `ctx_gguf` is not very clean. My suggestion would be to extend a map such as `tensors_offs` to include all the tensor info, and only use that from that point. The reason is that `ggml_get_tensor` is very inefficient since it does a linear search, which makes loading `O(N^2)`. Admittedly, the number of tensors is generally small enough that this is not a problem generally, but it is still not good.

Reply:
Thanks for your feedback. `llama_tensor_offset` moved inside `llama_model_loader`.

I see this approach more like the virtualization of a `ctx_gguf` than a hack.

I do not see what `clean` means from software design point of view. IMHO this solutions has:

1.`concerns`
- One can imagine the tensor data offset came from this virtual `gguf_context` instead of using `llama_tensor_offset`. But as the tensors data are accessible only once at the model loading, I think it's OK.

2. `benefits`
- all following calls to `ctx_ggux` and `ctx_meta` functions continue to work as expected
- No change to the existing logic, shipping the `split` feature as is is almost transparent to `llama.cpp`
- Supporting one or multiple files use the same logic, no additionnal if/else branch needed

@ggerganov I believe you will tell us what's the good direction to take and I will be happy to implement it, thanks

Reply:
I agree with @slaren's and @ngxson's PoVs - would prefer to avoid the merging into a virtual context and instead keep references to the existing contexts for each file. The main drawback of the proposed implementation is that we would have tensor offset logic in `llama.cpp` while this is already provided as information by the low-level `gguf/ggml` implementation. So the goal is to avoid having that implemented again in the user code

But other than that - this looks great 👍

Reply:
**TIMTOWTDI**, _vox populi_, the virtualization left 🤝. So I plan to have a `map<string,int>` to map the tensor name to the file index, LGTM. But how to deal with `get_tensor_name` with multiple `gguf_context` ?:
https://github.com/ggerganov/llama.cpp/blob/1943c0198125a0da1a200390e82cf461f9080d99/llama.cpp#L3021-L3024

Should I simply loop to all `gguf_context *` till the good tensor index ?
```c
    const char * get_tensor_name(int i) const {
        uint16_t idx = 0;
        gguf_context * gguf_ctx = ctx_ggufs.front().get();
        for (uint16_t tensor_idx = 0; tensor_idx < n_tensors; tensor_idx += gguf_get_n_tensors(gguf_ctx)) {
            if (i < tensor_idx + gguf_get_n_tensors(gguf_ctx)) {
                return gguf_get_tensor_name(gguf_ctx, i);
            }
            gguf_ctx = ctx_ggufs[++idx].get();
        }
        throw std::out_of_range("illegal tensor index");
    }
```

@slaren @ggerganov any better idea ?

Reply:
You don't need to keep all `ctx_gguf` - only the first one is necessary. This will be used to query KV values which are the same across all splits

You need to store the `ctx_ggml` from all `gguf_init_from_file`.

`get_tensor_name(int i)` and `get_tensor_meta(int i)` are trivial if you make an `std::vector` with all tensors from the all `ctx_ggml`. For `get_tensor_meta(const char * name)` we should probably iterate, unless you also add a map of `name -> i`

Reply:
Assuming that you already have the `map_name_to_i_file<tensor_name, i_file>` and that one file correspond to one `ctx_ggufs`, so `ctx_ggufs.size() == map_name_to_i_file.size()`

You can look up `i_file = map_name_to_i_file.at(tensor_name)`, then you can get the correct `ctx_gguf = ctx_ggufs[i_file]`

That's why earlier I propose having a map `tensor_name` to tuple of `ctx_gguf + file + mmap + ...`. But doing that will come with the cost of having to share the pointer of these objects, so having the `i_file` is also OK for now.

Reply:
In short, if you have one `ctx_gguf` and one `mmap` for each file, then you can use the same file index in the list of contexts and mmaps

Reply:
> To elaborate a bit more what I mean:

@slaren , thanks, are you suggesting 4 new fields in `llama_model_loader`: a `map<tensor_name, tensor_info>`, `vector<tensor_info>`, `vector<gguf_context*>`,  `vector<ggml_context*>`  with `tensor_info{gguf_context*, ggml_context*}` ?

Reply:
Not so many. Only keep the first `gguf_context` to access the KV, the rest can be freed immediately after loading the list of tensors. The map replaces the previous map. Only a new `std::vector<ggml_tensor*>` should be necessary for the functions that return the tensor or tensor name by index.

Reply:
> You don't need to keep all `ctx_gguf` - only the first one is necessary. This will be used to query KV values which are the same across all splits


OK, so you agree to keep track of the tensor data offset aside of `gguf_context` ? As I understood the opposite in the first message

Reply:
Yes, you do need to keep the list of `ggml_context` of all the files, since that's where the tensors are allocated. But only the first `gguf_context` is necessary.